core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -79,64 +97,114 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
+            fetchQueue.add(shareFetchPartitionData);
+        } finally {
+            lock.unlock();
+        }
+    }
 
-            replicaManager.fetchMessages(
-                fetchParams,
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry -> new Tuple2<>(entry.getKey(), entry.getValue())).collect(
-                        Collectors.toList())
-                ),
-                QuotaFactory.UnboundedQuota$.MODULE$,
-                responsePartitionData -> {
-                    List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
-                        responsePartitionData);
-                    Map<TopicIdPartition, ShareFetchResponseData.PartitionData> result = new HashMap<>();
-                    responseData.forEach(data -> {
-                        TopicIdPartition topicIdPartition = data._1;
-                        FetchPartitionData fetchPartitionData = data._2;
-
-                        partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition))
-                            .acquire(memberId, fetchPartitionData).whenComplete((acquiredRecords, throwable) -> {
-                                ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
-                                    .setPartitionIndex(topicIdPartition.partition());
-
-                                if (throwable != null) {
-                                    partitionData.setErrorCode(Errors.forException(throwable).code());
-                                } else {
-                                    // Maybe check if no records are acquired and we want to retry replica
-                                    // manager fetch. Depends on the share partition manager implementation,
-                                    // if we want parallel requests for the same share partition or not.
-                                    partitionData
-                                        .setPartitionIndex(topicIdPartition.partition())
-                                        .setRecords(fetchPartitionData.records)
-                                        .setErrorCode(fetchPartitionData.error.code())
-                                        .setAcquiredRecords(acquiredRecords)
-                                        .setAcknowledgeErrorCode(Errors.NONE.code());
-                                }
-                                result.put(topicIdPartition, partitionData);
-                            });
+    public boolean isFetchQueueEmpty() {
+        return fetchQueue.isEmpty();
+    }
+
+    public ShareFetchPartitionData pollFetchQueue() {
+        lock.lock();
+        try {
+            return fetchQueue.poll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void maybeProcessFetchQueue() {
+        try {
+            if (maybeAcquireProcessFetchQueueLock()) {
```

Comment:
I would say the if needs to be outside the try, because otherwise you will unconditionally release the lock, even if you didn't obtain it. There's another instance of this later on too.